### ISSUE TITLE
[feature] Enable a max parallelization flag to limit number of processes kicked off simultaneously

### ISSUE THREAD

**Issue by bryanjswift** [moonrepo/moon#477]
**Is your feature request related to a problem? Please describe.**

I have a monorepo with 30+ packages when I run a command like `moon run :shared-task` if there are not dependencies which bottleneck how many tasks can run then a `node` process will be started for each immediately this causes CPU thrashing or at least consumes all CPU meaning I can not do anything else while these jobs are in flight.

**Describe the solution you'd like**

I would like to be able to specify a maximum number of jobs that can be running in parallel at a given time. Either for a specific task or for the workspace in general.

**Describe alternatives you've considered**

Moving this specific task I'm working with into a script that I wrote that lets me control the parallelization for it. I've also considered making it a dependency of another task that does get limited by other dependencies so they wouldn't all kick off at the same time.

**Additional context**

The task I'm working with is applied to every project in my workspace and has no dependencies. In this case I'm using it to keep two files in sync (similar to what moon can do with `package.json` and `tsconfig.json`).

**Comment by milesj** [comment]
I'm currently not aware of a way to handle this, since the thread pool is entirely abstracted away behind Tokio (https://tokio.rs/).

I'll look into it however.

**Comment by bryanjswift** [comment]
> I'm currently not aware of a way to handle this, since the thread pool is entirely abstracted away behind Tokio (https://tokio.rs/).
> 
> I'll look into it however.

It looks like the Tokio runtime is currently implemented using [`#[tokio::main]`](https://docs.rs/tokio/latest/tokio/attr.main.html) by default I believe this creates an unbounded runtime. I suspect adding a `worker_threads` annotation to limit the number of simultaneous tasks spawned would be an "easy" first pass. To accept it as an option it may be necessary to use a runtime builder. https://docs.rs/tokio/latest/tokio/attr.main.html#set-number-of-worker-threads is the documentation I was looking at related to this.

**Comment by milesj** [comment]
I experimented with this here: https://github.com/moonrepo/moon/pull/506

It doesn't work how we expect it to, but I've talked the Tokio team about how to solve this and they gave me some ideas. Unfortunately it will take a bit of work and won't make the next release.

**Comment by bryanjswift** [comment]
> I experimented with this here: #506
> 
> It doesn't work how we expect it to, but I've talked the Tokio team about how to solve this and they gave me some ideas. Unfortunately it will take a bit of work and won't make the next release.

Copy. Thanks for the update, I appreciate it and look forward to see this land eventually.

**Comment by milesj** [comment]
This has landed in v0.22!

**Comment by bryanjswift** [comment]
> This has landed in v0.22!

Amazing! I'll be trying this out soon!
